miio/gateway/push_server.py

```diff
@@ -0,0 +1,297 @@
+import asyncio
+import calendar
+import datetime
+import logging
+import socket
+import struct
+from json import dumps
+from random import randint
+
+from ..protocol import Message, Utils
+
+_LOGGER = logging.getLogger(__name__)
+
+SERVER_PORT = 54321
+FAKE_DEVICE_ID = "120009025"
+FAKE_DEVICE_MODEL = "chuangmi.plug.v3"
+HELO_BYTES = bytes.fromhex(
+    "21310020ffffffffffffffffffffffffffffffffffffffffffffffffffffffff"
+)
+
+
+def calculated_token_enc(token):
+    token_bytes = bytes.fromhex(token)
+    encrypted_token = Utils.encrypt(token_bytes, token_bytes)
+    encrypted_token_hex = encrypted_token.hex()
+    return encrypted_token_hex[0:32]
+
+
+def construct_script(  # nosec
+    script_id,
+    action,
+    extra,
+    source_sid,
+    source_model,
+    target_id,
+    target_ip,
+    target_model,
+    token_enc,
+    message_id=0,
+    event=None,
+    command_extra="",
+    trigger_value=None,
+    trigger_token="",
+):
```

Comment:
For each device you need the same rather large function to properly format the message.
Therefore this function is defined here, but the device implements its own fucntion that calls this.
See `install_push_callbacks` function of the subdevice class: https://github.com/starkillerOG/python-miio/blob/a47301d6e2a96933f9de25e7974bb3bac26ca069/miio/gateway/devices/subdevice.py#L303-L350

But it is a bit of a mix between arguments of the device and the server

source_sid -> device
source_model -> device
token_enc -> device

target_id -> server
target_ip -> server
target_model -> server

script_id -> script specific
action -> script specific
extra -> script specific
event -> script specific
command_extra -> script specific
trigger_value -> script specific

I think it would be easiest to add it to the server, reducing the required arguments by 3, but it does not help a lot...

Reply:
I added it to the push server, somewhat simplifying things.